price-feeder/oracle/provider/crypto.go

```diff
@@ -347,65 +298,50 @@ func (p *CryptoProvider) messageReceived(messageType int, bz []byte, reconnectTi
 // When client receives an heartbeat message, it must respond back with the
 // public/respond-heartbeat method, using the same matching id,
 // within 5 seconds, or the connection will break.
-func (p *CryptoProvider) pong(heartbeatResp CryptoHeartbeatResponse, reconnectTicker *time.Ticker) {
-	reconnectTicker.Reset(cryptoReconnectTime)
-
+func (p *CryptoProvider) pong(heartbeatResp CryptoHeartbeatResponse) {
 	heartbeatReq := CryptoHeartbeatRequest{
 		ID:     heartbeatResp.ID,
 		Method: cryptoHeartbeatReqMethod,
 	}
 
-	if err := p.wsClient.WriteJSON(heartbeatReq); err != nil {
+	if err := p.wsc.SendJSON(heartbeatReq); err != nil {
 		p.logger.Err(err).Msg("could not send pong message back")
 	}
 }
 
-// ping to check websocket connection
-func (p *CryptoProvider) ping() error {
-	return p.wsClient.WriteMessage(websocket.PingMessage, ping)
-}
-
 func (p *CryptoProvider) setTickerPair(symbol string, tickerPair CryptoTicker) {
 	p.mtx.Lock()
 	defer p.mtx.Unlock()
 
-	price, err := coin.NewDecFromFloat(tickerPair.LatestTrade)
-	if err != nil {
-		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker price")
-		return
-	}
-	volume, err := coin.NewDecFromFloat(tickerPair.Volume)
+	tickerPrice, err := types.NewTickerPrice(
+		string(ProviderCrypto),
+		symbol,
+		tickerPair.LatestTrade,
+		tickerPair.Volume,
+	)
 	if err != nil {
-		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker volume")
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker")
 		return
 	}
 
-	p.tickers[symbol] = types.TickerPrice{
-		Price:  price,
-		Volume: volume,
-	}
+	p.tickers[symbol] = tickerPrice
 }
 
 func (p *CryptoProvider) setCandlePair(symbol string, candlePair CryptoCandle) {
 	p.mtx.Lock()
 	defer p.mtx.Unlock()
 
-	close, err := coin.NewDecFromFloat(candlePair.Close)
-	if err != nil {
-		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle close")
-		return
-	}
-	volume, err := coin.NewDecFromFloat(candlePair.Volume)
+	candle, err := types.NewCandlePrice(
+		string(ProviderCrypto),
+		symbol,
+		candlePair.Close,
+		candlePair.Volume,
+		candlePair.Timestamp,
```

Comment:
Do we not still have to convert this from seconds -> milliseconds? 

```suggestion
		SecondsToMilli(candlePair.Timestamp),
```

Reply:
good catch, for some reason I thought NewCandlePrice did that. https://github.com/umee-network/umee/pull/1554/commits/ef919f977998d74fd192688a008d607876dec610